hi-ml-histopathology/src/histopathology/models/deepmil.py

```diff
@@ -34,81 +46,152 @@ class BaseDeepMILModule(LightningModule):
 
+        self.loss_fn = self.get_loss()
+
         # Metrics Objects
         self.train_metrics = self.get_metrics()
         self.val_metrics = self.get_metrics()
         self.test_metrics = self.get_metrics()
 
+    @staticmethod
+    def get_checkpoint_downloader(ckpt_run_id: str, outputs_folder: Path) -> CheckpointDownloader:
+        downloader = CheckpointDownloader(
+            aml_workspace=get_workspace(),
+            run_id=ckpt_run_id,
+            checkpoint_filename=LAST_CHECKPOINT_FILE_NAME_WITH_SUFFIX,
+            download_dir=outputs_folder,
+            remote_checkpoint_dir=Path(f"{DEFAULT_AML_UPLOAD_DIR}/{CHECKPOINT_FOLDER}/")
+        )
+        downloader.download_checkpoint_if_necessary()
+        return downloader
+
+    def get_encoder(self, ckpt_run_id: Optional[str], outputs_folder: Optional[Path]) -> TileEncoder:
+        encoder: TileEncoder
+        if self.encoder_params.encoder_type == ImageNetEncoder.__name__:
+            encoder = ImageNetEncoder(
+                feature_extraction_model=resnet18,
+                tile_size=self.encoder_params.tile_size,
+                n_channels=self.encoder_params.n_channels,
+            )
+        elif self.encoder_params.encoder_type == ImageNetEncoder_Resnet50.__name__:
+            # Myronenko et al. 2021 uses Resnet50 CNN encoder
+            encoder = ImageNetEncoder_Resnet50(
+                feature_extraction_model=resnet50,
+                tile_size=self.encoder_params.tile_size,
+                n_channels=self.encoder_params.n_channels,
+            )
+
+        elif self.encoder_params.encoder_type == ImageNetSimCLREncoder.__name__:
+            encoder = ImageNetSimCLREncoder(
+                tile_size=self.encoder_params.tile_size, n_channels=self.encoder_params.n_channels
+            )
+
+        elif self.encoder_params.encoder_type == HistoSSLEncoder.__name__:
+            encoder = HistoSSLEncoder(
+                tile_size=self.encoder_params.tile_size, n_channels=self.encoder_params.n_channels
+            )
+
+        elif self.encoder_params.encoder_type == SSLEncoder.__name__:
+            assert ckpt_run_id and outputs_folder, "SSLEncoder requires ckpt_run_id and outputs_folder"
+            downloader = self.get_checkpoint_downloader(ckpt_run_id, outputs_folder)
+            encoder = SSLEncoder(
+                pl_checkpoint_path=downloader.local_checkpoint_path,
+                tile_size=self.encoder_params.tile_size,
+                n_channels=self.encoder_params.n_channels,
+            )
+        else:
+            raise ValueError(f"Unsupported encoder type: {self.encoder_params.encoder_type}")
+
+        if self.encoder_params.is_finetune:
+            for params in encoder.parameters():
+                params.requires_grad = True
+        else:
+            encoder.eval()
+        return encoder
+
+    def get_pooling_layer(self, pooling_params: PoolingParams) -> Tuple[nn.Module, int]:
```

Comment:
docu missing. Also, this function could live outside the class if I'm not mistaken, you'd only need `self.encoder.num_encoding` as an additonal arg?